Make the supervisor run on windows

mwrock opened the pull request:
This pr gets the supervisor running on windows with the following caveats:
1. The hooks are assumed to be powershell and use whatever `powershell.exe` is on the path
2. Processes are spawned as the current user and not as the specified `SVC` user/group
3. The process is always assumed to be healty and the implementation for `wait_for_pid` is a noop
4. The `chown` and `chmod` functionalities are no op on windows
5. ctrl+c (stopping the supervisor) will produce an error because it cannot send a kill signal

To test this I cloned https://github.com/habitat-sh/habitat-windows-package to a windows local `/hab/packages` note it is important to delete the `.git` folder after clone or else the version will not be read correctly.

thesentinels commented:
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

smurawski commented on components/core/src/os/filesystem/linux.rs:
```diff
@@ -15,11 +15,49 @@
+                                                       e)))
+        }
+    };
+    Ok(c_path.into_raw())
```
Is there a reason you are doing the into_raw here and return a raw pointer, vs returning the CString and doing the into_raw in the unsafe block where you do the from_raw?

mwrock replied:
yeah its returned to both the linux chown and chmod functions that need the raw pointer. This function allows this to be done in one place.

smurawski commented on components/core/src/package/install.rs:
```diff
@@ -299,13 +299,15 @@ impl PackageInstall {
-                    let package = try!(PackageIdent::from_str(id));
-                    if !package.fully_qualified() {
-                        return Err(Error::InvalidPackageIdent(package.to_string()));
+                if body.len() > 0 {
```
If there aren't deps, there shouldn't be a DEPS file, if the length of the body is 0, then we have a malformed package.

mwrock replied:
Ahhh. good to know. Thats what I get for hand rolling a package. I thought we always have a DEPS file. I have no idea why I assumed that.

mwrock replied:
Based on offline discusion this will remain.

smurawski commented on components/sup/src/command/shell.rs:
```diff
@@ -42,7 +42,7 @@ pub fn sh() -> Result<()> {
-    for path in try!(busybox_paths()).iter() {
+    for path in try!(interpreter_paths()).iter() {
         if !paths.is_empty() {
             paths.push(':');
```
This is a platform specific path separator and will break if multiple interpreter paths are specified on windows.  We moved to using `env::join_paths()` in https://github.com/habitat-sh/habitat/pull/1483/files#diff-b9b27ac18067cb62ebad852e9abc0b10R192 

Since this seems to be a pattern in a few places, maybe we should offer a function for this in core::os::env?

smurawski commented on components/sup/src/package/mod.rs:
```diff
@@ -106,7 +106,7 @@ impl Package {
         };
-        for path in try!(busybox_paths()).iter() {
+        for path in try!(interpreter_paths()).iter() {
             paths.push(':');
```
Same here with regard to path separator

mwrock commented on components/sup/src/util/mod.rs:
```diff
@@ -65,6 +69,41 @@ pub fn parse_ip_port_with_defaults(s: Option<&str>,
+    use std::os::unix::process::CommandExt;
+    let uid = os::users::get_uid_by_name(user);
+    let gid = os::users::get_gid_by_name(group);
+    if let None = uid {
```
This was a pure cut/paste but I like your idea!

mwrock commented:
I've incorporated the cleanup recomendations. I think it is solid except the use of `to_owned` in the newly extracted path appending helper method feels like I might be doing the wrong thing so please suggest changes if necessary.

smurawski commented on components/sup/src/command/shell.rs:
```diff
@@ -15,13 +15,14 @@
 
 use error::{Error, Result};
-use util::path::busybox_paths;
+use util::path::append_interpreter_and_path;
```
Rust style tends to prefer not directly importing funtions, and import the module then module qualify the function.

https://aturon.github.io/style/imports.html#prefer-fully-importing-types/traits-while-module-qualifying-functions.

smurawski commented on components/sup/src/package/mod.rs:
```diff
@@ -35,7 +35,7 @@ use error::{Error, Result, SupError};
 use service_config::ServiceConfig;
 use supervisor::Supervisor;
-use util::path::busybox_paths;
+use util::path::append_interpreter_and_path;
```
Same thing about not directly importing the function

smurawski commented on components/sup/src/supervisor.rs:
```diff
@@ -23,17 +23,17 @@ use std::fs::{self, File};
 
 use error::{Result, Error};
-use util::signals;
+use util::{create_command, signals};
```
e.g. here, signals is a module that is being imported vs. create_command which is a function.

smurawski commented on components/sup/src/util/path.rs:
```diff
@@ -123,6 +124,24 @@ pub fn busybox_paths() -> Result<Vec<PathBuf>> {
+    let mut paths = try!(interpreter_paths()).to_owned();
+    orig_paths.append(&mut paths);
+    if let Some(val) = env::var_os("PATH") {
+        let mut os_paths = env::split_paths(&val).collect::<Vec<PathBuf>>().to_owned();
```
Does the `to_owned()` actually solve any problems?  Just running a build and the test cycle doesn't seem to break anything without it.

mwrock replied:
heh. you are right. Probably another case where it fixed something before the rest of my keyboard mashing that fixed it.

mwrock commented:
ok thanks @smurawski just made the changes.

smurawski commented:
The travis test failure is due to a file that's missing (which is in master).

![](https://m.popkey.co/0d3902/5mVaz.gif)

mwrock commented:
I rebased this and cleaned up conflicts. The windows `build.ps1` now pulls down the zmq windows binaries from nuget. After I got the build to build zmq on windows I discovered that rust-zmq is not currently windows compatible. This is a known issue and one of the maintainers is aware of it. There is a PR that resolves this and the maintainer is looking to merge it. (9 days ago). So I forked the PR branch into our org for us to use until its merged in upstream.

Additionally this PR added some no-op behavior for windows signal handling and now includes the windows error codes for `TIMEOUT` and `WOULDBLOCK` on the inbound supervisoe thread so it does not report errors on each loop iteration when there is no data on the socket.

This passes tests here  and runs on my wuindows environment just as demoed last week. I've kept all of these changes in the last commit so they can be reviewed easily.

smurawski commented on build.ps1:
```diff
@@ -77,6 +82,10 @@ function Invoke-Configure {
         choco install habitat_native_dependencies --confirm -s https://www.myget.org/F/habitat/api/v2  --allowemptychecksums
     }
 
+    choco install libzmq_vc120 --version 4.2.3 --confirm -s https://www.nuget.org/api/v2/ --allowemptychecksums
```
This is probably ok for now, but everything else (including the rust bits) is binding against the vc140 runtime.

smurawski commented on components/builder-depot/Cargo.toml:
```diff
@@ -64,7 +64,7 @@ git = "https://github.com/habitat-sh/urlencoded.git"
 
 [dependencies.zmq]
-git = "https://github.com/erickt/rust-zmq.git"
+git = "https://github.com/habitat-sh/rust-zmq.git"
```
We can also pin to a specific branch, tag, or revision, or maybe add a comment to explain why we are currently binding to a fork here and what the conditions are to go back to the standard crate.

http://doc.crates.io/specifying-dependencies.html#specifying-dependencies-from-git-repositories

mwrock replied:
currently the PR was the master branch of the submitter's repo so putting it here seemed safest but aree a comment will be good. Hopefully this repo won't live long.

mwrock replied:
I tagged our repo and added that to the dependency declarations.

smurawski commented on components/sup/src/package/hooks.rs:
```diff
@@ -16,18 +16,17 @@ use std::fmt;
 use util::convert;
 use util::handlebars_helpers;
 use util::users as hab_users;
+use util::create_command;
```
Rust import guidelines lean against importing individual functions.  This could be `use util` with `util::create_comand` being called below.

mwrock commented on appveyor.yml:
```diff
@@ -21,7 +21,7 @@ environment:
     - CARGO_HOME: "c:\\cargo"
       hab_build_action: "test;build"
-      hab_components: "core;http-client;builder-protocol;builder-depot-client;common;win-users"
+      hab_components: "core;http-client;builder-protocol;builder-depot-client;common;win-users;sup"
```
ideally yes but it was intermittently failing 1 `cargo test` and should address that separately.

mwrock commented:
@thesentinels approve

thesentinels commented:
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

thesentinels commented:
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/178964570).

thesentinels commented:
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/178964570).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

![gif-keyboard-3280869874741411265](https://cloud.githubusercontent.com/assets/4304/17755674/1e577b82-6490-11e6-96b7-1663d283c824.gif)
